src/Simplex/Messaging/Agent/Client.hs

```diff
@@ -535,35 +530,43 @@ getSMPServerClient c@AgentClient {active, smpClients, msgQ} tSess@(userId, srv,
         notifySub connId cmd = atomically $ writeTBQueue (subQ c) ("", connId, APC (sAEntity @e) cmd)
 
 reconnectServer :: AgentMonad m => AgentClient -> SMPTransportSession -> m ()
-reconnectServer c tSess = newAsyncAction tryReconnectSMPClient $ reconnections c
+reconnectServer c tSess =
+  atomically setup >>= \case
+    Left reserved -> do
+      a <- async $ tryReconnectSMPClient `E.catchAny` const (atomically cleanup)
+      atomically (takeTMVar reserved >> putTMVar reserved (Just a))
+    Right running -> atomically $ tryReadTMVar running >>= mapM_ (maybe retry waitSTM)
   where
-    tryReconnectSMPClient aId = do
+    setup = getTSessVar tSess (reconnections c) >>= bimapM (\v -> v <$ putTMVar v Nothing) pure
+    tryReconnectSMPClient = do
       ri <- asks $ reconnectInterval . config
       timeoutCounts <- newTVarIO 0
-      withRetryIntervalCount ri $ \n _ loop ->
-        reconnectSMPClient n timeoutCounts c tSess `catchAgentError` const loop
-      atomically . removeAsyncAction aId $ reconnections c
-
-reconnectSMPClient :: forall m. AgentMonad m => Int -> TVar Int -> AgentClient -> SMPTransportSession -> m ()
-reconnectSMPClient n tc c tSess@(_, srv, _) = do
-  ts <- liftIO getCurrentTime
-  let label = unwords ["reconnect", show n, show ts]
-  withLockMap_ (reconnectLocks c) tSess label $ do
-    qs <- atomically (RQ.getSessQueues tSess $ pendingSubs c)
-    NetworkConfig {tcpTimeout} <- readTVarIO $ useNetworkConfig c
-    -- this allows 3x of timeout per batch of subscription (90 queues per batch empirically)
-    let t = (length qs `div` 90 + 1) * tcpTimeout * 3
-    t `timeout` mapM_ resubscribe (L.nonEmpty qs) >>= \case
-      Just _ -> atomically $ writeTVar tc 0
-      Nothing -> do
-        tc' <- atomically $ stateTVar tc $ \i -> (i + 1, i + 1)
-        maxTC <- asks $ maxSubscriptionTimeouts . config
-        let err = if tc' >= maxTC then CRITICAL True else INTERNAL
-            msg = show tc' <> " consecutive subscription timeouts: " <> show (length qs) <> " queues, transport session: " <> show tSess
-        atomically $ writeTBQueue (subQ c) ("", "", APC SAEConn $ ERR $ err msg)
+      withRetryIntervalCount ri $ \_ _ loop -> do
+        pending <- atomically $
+          RQ.getSessQueues tSess (pendingSubs c) >>= \case
+            [] -> Nothing <$ cleanup
+            q : qs -> pure $ Just (q :| qs)
+        forM_ pending $ \qs -> do
+          reconnectSMPClient timeoutCounts c tSess qs `catchAgentError` \_ -> pure ()
+          loop
```

Comment:
I am still thinking about growing timeout here, maybe we can somehow end it sooner, not sure.